Jenkinsfile

```diff
@@ -1,117 +1,151 @@
+        Note that a release build will not promote the build, but rather will promote the most recent milestone build.
+        '''
+    )
+
+    choice(
+      name: 'TARGET_PLATFORM',
+      choices: [ 'eclipse_2023_03-xtext_2_31_0', 'eclipse_2023_09-xtext_nightly' ],
+      description: '''
+        Choose the named target platform against which to compile and test.
+        This is relevant only for nightly and milestone builds.
+        '''
+    )
+
+    booleanParam(
+      name: 'ECLIPSE_SIGN',
+      defaultValue: true,
+      description: '''
+        Choose whether or not the bundles will be signed.
+        This is relevant only for nightly and milestone builds.
+      '''
+    )
+
+    booleanParam(
+      name: 'PROMOTE',
+      defaultValue: true,
+      description: 'Whether to promote the build to the download server.'
+    )
+  }
+
+  stages {
+    stage('Display Parameters') {
+      steps {
+        script {
+          env.BUILD_TYPE = params.BUILD_TYPE
+          env.TARGET_PLATFORM = params.TARGET_PLATFORM
+
+          if (env.BRANCH_NAME == 'master') {
+            // Only sign the master branch.
+            env.ECLIPSE_SIGN = params.ECLIPSE_SIGN
+          } else {
+            env.ECLIPSE_SIGN =  false
+          }
+
+            // Only promote signed builds.
+          env.PROMOTE = params.PROMOTE && (env.ECLIPSE_SIGN == 'true')
         }
-        } catch(e) {
-            currentBuild.result = 'FAILED'
-            throw e
-        } finally {
-            postAlways()
+        echo """
+BUILD_TIMESTAMP=${env.BUILD_TIMESTAMP}
+BUILD_TYPE=${env.BUILD_TYPE}
+TARGET_PLATFORM=${env.TARGET_PLATFORM}
+ECLIPSE_SIGN=${env.ECLIPSE_SIGN}
+PROMOTE=${env.PROMOTE}
+BRANCH_NAME=${env.BRANCH_NAME}
+"""
+      }
+    }
+
+    stage('Build') {
+      steps {
+        sshagent(['projects-storage.eclipse.org-bot-ssh']) {
+          wrap([$class: 'Xvnc', useXauthority: true]) {
+            dir('.') {
+              sh '''
+                if [[ $PROMOTE != true ]]; then
+                  promotion_argument='-P!promote'
+                fi
+                mvn  \
+                  --fail-at-end \
+                  --no-transfer-progress \
+                  --update-snapshots \
+                  $promotion_argument \
+                  -Dmaven.repo.local=xpect-local-maven-repository \
+                  -Dmaven.artifact.threads=16 \
+                  -Dbuild.id=${BUILD_TIMESTAMP} \
+                  -Dgit.commit=$GIT_COMMIT \
+                  -Declipsesign=${ECLIPSE_SIGN} \
+                  -Dbuild.type=$BUILD_TYPE \
+                  -Dtycho-version=4.0.3 \
+                  -Dtarget-platform=${TARGET_PLATFORM} \
+                  -Dorg.eclipse.justj.p2.manager.build.url=$JOB_URL \
+                  -Dorg.eclipse.justj.p2.manager.relative=$PUBLISH_LOCATION \
+                  clean \
+                  verify
+                '''
+            }
+          }
         }
+      }
     }
-}
+  }
 
+  post {
```

Comment:
i assume we want to keep matrix here

Reply:
I don't know what that means.

Reply:
Ah, all that postAlways() stuff that I don't understand.  I assume we want that in, but that's just copy and paste, right?  And then called from the "always{}" block.    I don't think I could test this on my own instance...

Reply:
exactly.
we use eclipse chat for build notificiations
see
https://chat.eclipse.org/#/room/#modeling-xpect-build:matrix.eclipse.org

Reply:
That looks cool.  Shall I try to amend my commit and force push it to add this?